src/face_op.cpp

```diff
@@ -192,38 +247,120 @@ void Manifold::Impl::Face2Tri(const Vec<int>& faceEdge,
- * must be referenced the same number of times as a startVert and endVert.
- */
-PolygonsIdx Manifold::Impl::Face2Polygons(VecView<Halfedge>::IterC start,
-                                          VecView<Halfedge>::IterC end,
-                                          mat2x3 projection) const {
-  std::multimap<int, int> vert_edge;
-  for (auto edge = start; edge != end; ++edge) {
-    vert_edge.emplace(
-        std::make_pair(edge->startVert, static_cast<int>(edge - start)));
+void Manifold::Impl::FlattenFaces() {
+  Vec<uint64_t> edgeFace(halfedge_.size());
+  constexpr uint64_t remove = std::numeric_limits<uint64_t>::max();
+  const size_t numTri = NumTri();
+  const auto policy = autoPolicy(numTri);
+
+  std::vector<std::atomic<int>> vertDegree(NumVert());
+  for_each(policy, vertDegree.begin(), vertDegree.end(),
+           [](auto& v) { v.store(0); });
+
+  for_each_n(policy, countAt(0_uz), numTri,
+             [&edgeFace, &vertDegree, remove, this](size_t tri) {
+               for (const int i : {0, 1, 2}) {
+                 const int edge = 3 * tri + i;
+                 const int pair = halfedge_[edge].pairedHalfedge;
+                 const auto& ref = meshRelation_.triRef[tri];
+                 if (ref.SameFace(meshRelation_.triRef[pair / 3])) {
+                   edgeFace[edge] = remove;
+                 } else {
+                   edgeFace[edge] = (static_cast<uint64_t>(ref.meshID) << 32) +
+                                    static_cast<uint64_t>(ref.faceID);
+                   ++vertDegree[halfedge_[edge].startVert];
+                 }
+               }
+             });
+
+  Vec<size_t> newHalf2Old(halfedge_.size());
+  sequence(newHalf2Old.begin(), newHalf2Old.end());
+  stable_sort(
+      newHalf2Old.begin(), newHalf2Old.end(),
+      [&edgeFace](size_t a, size_t b) { return edgeFace[a] < edgeFace[b]; });
+  newHalf2Old.resize(std::find_if(countAt(0_uz), countAt(halfedge_.size()),
+                                  [&](const size_t i) {
+                                    return edgeFace[newHalf2Old[i]] == remove;
+                                  }) -
+                     countAt(0_uz));
+
+  Vec<Halfedge> newHalfedge(newHalf2Old.size());
+  for_each_n(policy, countAt(0_uz), newHalf2Old.size(),
+             [&](size_t i) { newHalfedge[i] = halfedge_[newHalf2Old[i]]; });
+
+  Vec<int> faceEdge(1, 0);
+  for (size_t i = 1; i < newHalf2Old.size(); ++i) {
+    if (edgeFace[newHalf2Old[i]] != edgeFace[newHalf2Old[i - 1]]) {
+      faceEdge.push_back(i);
+    }
   }
+  const int numFace = faceEdge.size();
+  faceEdge.push_back(newHalf2Old.size());
 
-  PolygonsIdx polys;
-  int startEdge = 0;
-  int thisEdge = startEdge;
-  while (1) {
-    if (thisEdge == startEdge) {
-      if (vert_edge.empty()) break;
-      startEdge = vert_edge.begin()->second;
-      thisEdge = startEdge;
-      polys.push_back({});
+  if (numFace < 4) {
+    MarkFailure(Error::NoError);
```

Comment:
Should we change the name of this function? `Done` or something? Otherwise it is a bit weird to mark something as failure even though it has no error.

Reply:
Yeah, I was thinking the same thing. MarkEmpty sounds good.